storage/client_test.go

```diff
@@ -1506,6 +1506,226 @@ func TestRetryReadStallEmulated(t *testing.T) {
+				instructions: map[string][]string{
+					"storage.objects.insert": {"stall-for-10s-after-3072K"},
+				},
+				wantDuration: 10 * time.Second,
+			},
+			{
+				name: "stall-on-first-chunk-twice-with-chunk-transfer-timeout-zero",
+				instructions: map[string][]string{
+					"storage.objects.insert": {"stall-for-10s-after-1024K", "stall-for-10s-after-1024K"},
+				},
+				wantDuration: 10 * time.Second,
+			},
+		}
+
+		for _, tc := range tests {
+			t.Run(tc.name, func(t *testing.T) {
+				testID := createRetryTest(t, client, tc.instructions)
+				ctx := callctx.SetHeaders(ctx, "x-retry-test-id", testID)
+
+				prefix := time.Now().Nanosecond()
+				want := &ObjectAttrs{
+					Bucket:     bucket,
+					Name:       fmt.Sprintf("%d-object-%d", prefix, time.Now().Nanosecond()),
+					Generation: defaultGen,
+				}
+
+				var gotAttrs *ObjectAttrs
+				params := &openWriterParams{
+					attrs:     want,
+					bucket:    bucket,
+					chunkSize: chunkSize,
+					ctx:       ctx,
+					donec:     make(chan struct{}),
+					setError:  func(_ error) {}, // no-op
+					progress:  func(_ int64) {}, // no-op
+					setObj:    func(o *ObjectAttrs) { gotAttrs = o },
+				}
+
+				startTime := time.Now()
+				pw, err := client.OpenWriter(params)
+				if err != nil {
+					t.Fatalf("failed to open writer: %v", err)
+				}
+				buffer := bytes.Repeat([]byte("A"), fileSize)
+				if _, err := pw.Write(buffer); err != nil {
+					t.Fatalf("failed to populate test data: %v", err)
+				}
+				if err := pw.Close(); err != nil {
+					t.Fatalf("closing object: %v", err)
+				}
+				endTime := time.Now()
+				select {
+				case <-params.donec:
+				}
+				if gotAttrs == nil {
+					t.Fatalf("Writer finished, but resulting object wasn't set")
+				}
+				if diff := cmp.Diff(gotAttrs.Name, want.Name); diff != "" {
+					t.Fatalf("Resulting object name: got(-),want(+):\n%s", diff)
+				}
+				if endTime.Sub(startTime) < tc.wantDuration {
+					t.Errorf("write took %v, want >= %v", endTime.Sub(startTime), tc.wantDuration)
+				}
+
+				r, err := veneerClient.Bucket(bucket).Object(want.Name).NewReader(ctx)
+				if err != nil {
+					t.Fatalf("opening reading: %v", err)
+				}
+				wantLen := len(buffer)
+				got := make([]byte, wantLen)
+				n, err := r.Read(got)
+				if n != wantLen {
+					t.Fatalf("expected to read %d bytes, but got %d", wantLen, n)
+				}
+				if diff := cmp.Diff(got, buffer); diff != "" {
+					t.Fatalf("checking written content: got(-),want(+):\n%s", diff)
+				}
+			})
+		}
+	})
+}
+
+// In resumable uploads, the chunk encounter a stall. The chunkTransferTimeout will then cancel the request and resend it.
+// The second request is expected to succeed. This allows the upload to finish earlier than if it had waited for the entire stall
+// duration, as the chunkTransferTimeout prevents unnecessary delays.
+func TestRetryWriteReqStallWithNonZeroChunkTransferTimeoutEmulated(t *testing.T) {
```

Comment:
nit: I'd just call this: TestWriterChunkTransferTimeoutEmulated.

You can also combine both tests; just separate out into cases which are expected to succeed and those which are expected to fail.

Reply:
Done